### PR TITLE
Test dill

### PR DESCRIPTION
**Pull request by albertvillanova**
Regression test for future releases of `dill`.

Related to #4379.

### CODE REVIEW

**Comment by anivegesana** [comment]
I should point out that the hash will be the same if computed twice with the same code on the same version of dill (after adding huggingface's code that removes line numbers and file names, and sorts globals.) My changes in dill 0.3.5 and ones that I will make in 0.3.6 will result in different pickles than the ones dill 0.3.4 was making. This should still be fine for caching.

File: tests/test_fingerprint.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+from datasets.fingerprint import Hasher
+
+
+def test_dill():
+    # AttributeError: module 'dill._dill' has no attribute 'stack'
+    hasher = Hasher()
+    hasher.update(lambda x: x)
```
Comment (lhoestq) [inline tests/test_fingerprint.py:7]: I think you can just check this
```suggestion
    Hasher.hash(lambda x: x)
```

**Review by lhoestq — COMMENTED** [review]
Thanks !
I think you can have this test in `test_caching.py` with the other tests on hashing :)

**Comment by albertvillanova** [comment]
Just some comments @lhoestq:

The best practice for testing is to have a `test_<filename>.py` for each `<filename>.py`. Therefore in order to have the filenames aligned, I would propose:
- either renaming `fingerprint.py` to `caching.py`
- or renaming `test_caching.py` to `test_fingerprint.py`

On the other hand, my idea when implementing this test was not to test all the functionalities of the `Hasher`, but just to have a regression test that fails if dill version is > 0.3.4 and the pin in our `setup.py` is not present. Just recall that we had no failing test in our CI when the issue with dill was found on `transformers`.

The objective of this PR is just to have a regression test for that case: I tested and I got `AttributeError: module 'dill._dill' has no attribute 'stack'`

There was no equivalent test in `test_caching.py` because the our CI did not fail before pinning dill.

**Pull request merged**